### DISCUSSION TITLE
Include Dynamic Node Ports in JSON

### DISCUSSION THREAD

**Discussion by ghost** [chrisjpatty/flume#61] [Q&A]
I'm saving the node configuration as json on the server. With dynamic node ports that's not possible. Maybe it should be possible to refer to functions via a string and flume calls them like this
```javascript
["functionName"](arguments)
```

**Comment by chrisjpatty** [comment]
Hey @patrickap-bkklinde! Trying to parse functions in and out of JSON can have some serious security implications that are probably out of the scope of this library, but that being said, this is something you could implement on your own end if you're needing it. Like you've seen, the configuration objects are just JSON objects with the exception of the dynamic field objects. You can create your own function that converts the configuration object to a string and stores a reference to a function name in the place of the function. Then when you retrieve your configuration JSON from your database, you can rehydrate it with the functions you need based on their names. Does that work for you in this case?

**Reply by ghost** [reply]
Hi @chrisjpatty , first of all thank you for your answer. I agree with you that this involves security risks. My idea was not to stringify and parse functions or to call some code via `eval`, I was thinking more of an API so that I can define such functions in the scope of Flume and Flume itself only saves a string to this function in the configuration JSON. When it's loaded into Flume it could just find the function. Does that pose the same risk?

**Reply by chrisjpatty** [reply]
@patrickap-bkklinde Maybe I'm not fully understanding what you're trying to do. Flume wouldn't necessarily know how to map function names to functions you provide unless there was a separate structure containing that mapping.

**Reply by ghost** [reply]
@chrisjpatty Sorry, i will try to explain it in detail. The advantage of my idea: I can save static string references of dynamic node functions in the flume config / json. Here is an example.

```javascript
// something like that could be part of the flume api
const storeFunctions = (...fns) => {
  const store = Object.fromEntries(fns.map(fn =>([
    fn.name, fn
  ])))

  return {
    call: (str) => store[str]()
  }
}

// define functions / conditions for my dynamic nodes
const sayHello = () => alert('Hello');
const sayWorld = () => alert('World');

// store the functions so they can be called later by flume
const storedFunctions = storeFunctions(
  sayHello,
  sayWorld
)

// flume could read it's configuration and call them like this
storedFunctions.call('sayHello')
storedFunctions.call('sayWorld')
```

This makes it possible to easily save the json on the server even with dynamic nodes and load it back into flume. The configuration json does not contain the function itself, only the name as string.

**Reply by ghost** [reply]
@chrisjpatty Or to be more precise (the data model is not entirely correct, but i think my idea will become clear):

```javascript
// ----- current -----
config.
  addPortType({...})
  addNodeType({
    ...,
    inputs: ports => data => {
      ...
      return [...];
    },
    ...,
  })

// generates a json
{
  ...,
  inputs: function() {...}
  ...,
}

// ----- suggestion -----
config.
  addPortType({...})
  addNodeType({
    ...,
    inputs: storedFunctions.call('functionName')
    ...,
  })

// generates a json
{
  ...,
  inputs: 'functionName'
  ...,
}
```

**Reply by ghost** [reply]
Hello @chrisjpatty, I just wanted to ask if there is any news on this? Wouldn't that be useful somehow?
